### ISSUE TITLE
More idiomatic code instead of impl_* macros

### ISSUE THREAD

**Issue by dr-orlovsky** [rust-bitcoin/rust-bitcoin#352]
I use rust-bitcoin a lot as a dependency and do need to create new wrapper types (like we did for the hash types) a lot. However, in order to implement them, I need a lot of different internal macro to be made exportable. While this is an option, there is another option, we can consider. For the boilerplate impl code rust allows to use generics instead of macros, which, I assume, is more idiomatic. So for instance, instead of using `impl_index_newtype!` https://github.com/rust-bitcoin/rust-bitcoin/blob/885fc39f4881f26929a0c3303cfccce9497ba1e4/src/internal_macros.rs#L175 for each type and making it exportable we can create just a single implementation which will cover all the cases, like this one:

```rust
#[derive(Clone, PartialEq, Eq)]
pub struct Wrapper<T, Z>(T, PhantomData<Z>);

impl<T, Z> From<T> for Wrapper<T, Z> {
    #[inline]
    fn from(x: T) -> Self { Self(x, PhantomData::default()) }
}

impl<T, U, Z> Index<Range<usize>> for Wrapper<T, Z> 
    where T: Index<Range<usize>, Output=[U]> 
{
    type Output = [U];

    #[inline]
    fn index(&self, index: Range<usize>) -> &[U] {
        &self.0[index]
    }
}

impl<T, U, Z> Index<RangeTo<usize>> for Wrapper<T, Z> 
    where T: Index<RangeTo<usize>, Output=[U]> 
{
    type Output = [U];

    #[inline]
    fn index(&self, index: RangeTo<usize>) -> &[U] {
        &self.0[index]
    }
}

impl<T, U, Z> Index<RangeFrom<usize>> for Wrapper<T, Z> 
    where T: Index<RangeFrom<usize>, Output=[U]> 
{
    type Output = [U];

    #[inline]
    fn index(&self, index: RangeFrom<usize>) -> &[U] {
        &self.0[index]
    }
}

impl<T, U, Z> Index<RangeFull> for Wrapper<T, Z> 
    where T: Index<RangeFull, Output=[U]> 
{
    type Output = [U];

    #[inline]
    fn index(&self, _: RangeFull) -> &[U] {
        &self.0[..]
    }
}
```

...and no macro use for each new type is required! To define a new wrapper type, like for hash type Txid, one need just to write

```rust
struct _TxidPhantom;
pub type Txid = Wrapper<sha256::Hash, _TxidPhantom>;
```

..and it will give the whole Txid functionality w/o calling to macro or doing impl's. 
The same can be done not just for a Wrapper types, but in many other cases where we use the macros. So I do propose to write more idiomatic rust code and get rid from internal macros wherever is possible – I am ready to work on this change, since I have to do it otherwise somewhere on top of rust-bitcoin lib.

**Comment by apoelstra** [comment]
Very cool. concept ACK. cc @TheBlueMatt thoughts?

**Comment by elichai** [comment]
I don't quite like it. on one hand you use type aliasing, on another you use a phantom type to prevent having these types interchangeable. I personally prefer implementing `Deref`+`DerefMut` and then deriving anything else (partialEq,Eq,Clone etc.) 

the deref black magic should give the `Index` and all the rest of the operations for free

**Comment by thomaseizinger** [comment]
From a idiomatic Rust code perspective, I'd say a custom derive is probably the best solution for the boilerplate of forwarding trait implementations to an inner type.

**Comment by elichai** [comment]
> From a idiomatic Rust code perspective, I'd say a custom derive is probably the best solution for the boilerplate of forwarding trait implementations to an inner type.

That sounds like the best solution IMHO, Even though I like abusing Deref there's a limit for that :)
(and that can be a generic crate for implementing index)

**Comment by thomaseizinger** [comment]
The derive-more crate offers such a custom-derive: https://jeltef.github.io/derive_more/derive_more/index_op.html

**Comment by elichai** [comment]
> The derive-more crate offers such a custom-derive: https://jeltef.github.io/derive_more/derive_more/index_op.html

They also use proc_macro so you can choose a field. this means it doesn't works on 1.22.0

I can probably write something specific to Index in less than a day

**Comment by thomaseizinger** [comment]
> They also use proc_macro so you can choose a field. this means it doesn't works on 1.22.0

Uff, that's a bummer :(

**Comment by dr-orlovsky** [comment]
My original point was mostly about avoiding use of macro and preferring open implementations. Derive seems to be the best strategy, so I will stick to it (to the extent which is allowed for the MSRC). But the question is would it be better to have something like

```rust
#[derive(Index, Clone, PartialEq, Eq)]
pub struct Wrapper<T, Z>(T, PhantomData<Z>);

impl<T, Z> From<T> for Wrapper<T, Z> {
    #[inline]
    fn from(x: T) -> Self { Self(x, PhantomData::default()) }
}
```

comparing to a hundreds of lines of macro code as of today...

**Comment by elichai** [comment]
> My original point was mostly about avoiding use of macro and preferring open implementations. Derive seems to be the best strategy, so I will stick to it (to the extent which is allowed for the MSRC). But the question is would it be better to have something like
> 
> ```rust
> #[derive(Index, Clone, PartialEq, Eq)]
> pub struct Wrapper<T, Z>(T, PhantomData<Z>);
> 
> impl<T, Z> From<T> for Wrapper<T, Z> {
>     #[inline]
>     fn from(x: T) -> Self { Self(x, PhantomData::default()) }
> }
> ```
> 
> comparing to a hundreds of lines of macro code as of today...

I thought the problem is mostly the `Index` implementations? if we write a `Index` implementation that assume a single field we can easily add `From` to it

**Comment by dr-orlovsky** [comment]
No, the main problem is plenty of nested macros for impl_..., which are really hard to debug and introspect. Right now at least quarter of all rust-bitcoin code is generated by macros, and I am looking for a way to avoid them as much as possible and convert that into a normal rust code without repeating it.

**Comment by thomaseizinger** [comment]
> No, the main problem is plenty of nested macros for impl_..., which are really hard to debug and introspect. Right now at least quarter of all rust-bitcoin code is generated by macros, and I am looking for a way to avoid them as much as possible and convert that into a normal rust code without repeating it.

That sounds like custom-derives are your solution :)
We just need to find a good split but in general, you don't need to look at the generated code of a custom-derive. So far, the main usecase mentioned here is forwarding implementations of certain traits to the inner type of a wrapper. Is there more code you would like to have generated?

**Comment by thomaseizinger** [comment]
IIRC, there are ways of programmatically detecting the version of rustc that is in use. We could also send a PR to the derive-more crate and lower the MSRV by gating code that would need a newer version. That would remove the need for maintaining another crate that just generates boilerplate code. derive-more is already heavily feature-gated so they might be open to this :)

**Comment by dr-orlovsky** [comment]
Basically, it will be nice to re-do everything in https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/internal_macros.rs, namely automatic implementations for:
* Encodable/Decodable
* From, Index, custom PartialEq/PartialOrder
* Serde
* FromStr / FromHex
* standard trait impls for enums (Debug, Display, FromStr, Serde)

Yes, this all sounds like a case for custom derive. The only question is what can be done with regards to our current MSRV...

**Comment by thomaseizinger** [comment]
You would like to access them for the `impl_hashnewtype` macro (I don't recall the actual name) right?. What was the reason for separating `bitcoin_hashes` out into a 2nd crate @apoelstra ?
It doesn't make the macros go away, but it would keep them internal if bitcoin_hashes were part of rust-bitcoin.

**Comment by dr-orlovsky** [comment]
No, it have nothing to do with `impl_hashnewtype`, which is already working fine. It's more about making the code better. Also, I need to build a number of abstractions outside of rust-bitcoin (in https://github.com/lnp-bp/rust-lnpbp) and would like not to export the internal macros – and not re-implement many things which can be doe within rust-bitcoin lib itself

**Comment by elichai** [comment]
> Basically, it will be nice to re-do everything in https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/internal_macros.rs, namely automatic implementations for:
> 
>     * Encodable/Decodable
> 
>     * From, Index, custom PartialEq/PartialOrder
> 
>     * Serde
> 
>     * FromStr / FromHex
> 
>     * standard trait impls for enums (Debug, Display, FromStr, Serde)
> 
> 
> Yes, this all sounds like a case for custom derive. The only question is what can be done with regards to our current MSRV...

i'll look into this

**Comment by elichai** [comment]
arghhh all the `Index` implementations could get drilled down to https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=babc2033f45e276c230fe9297ba2bd47 but only from rust 1.28 (When `SliceIndex` was stabilized)

**Comment by apoelstra** [comment]
@thomaseizinger `bitcoin_hashes` is a separate crate so that I can do signature stuff without dragging in all of rust-bitcoin.

**Comment by apoelstra** [comment]
and I have a vague plan to make `secp256k1` depend on `bitcoin_hashes`

**Comment by dr-orlovsky** [comment]
Just for the reference here is my take on the `Wrapper` type which I ended using myself to do a custom type derivation (not hashtype based, more generic): https://github.com/LNP-BP/rust-lnpbp/blob/master/src/common/wrapper.rs

**Comment by cloudhead** [comment]
Note that a lot of the ideas here are now possible since https://github.com/rust-bitcoin/rust-bitcoin/commit/6138084c5bce4b3013fffc6d6341399d5c71f737

**Comment by tcharding** [comment]
Hey @dr-orlovsky do #805 and #870 combined resolve this issue in your opinion, or are there more macros in `internal_macros.rs` that you would like to see removed?

**Comment by dr-orlovsky** [comment]
This is a good work, thank you! There is also `impl_bytes_newtype`, which I am not sure we can address...

BTW, I ended up having my own solution with `Wrapper` derive macro here: https://docs.rs/amplify_derive/2.11.2/amplify_derive/derive.Wrapper.html
@Kixunil is this what you was looking for?

**Comment by tcharding** [comment]
Is this still an issue or are we happy with our use of declarative macros in `internal_macros`?

**Comment by Kixunil** [comment]
Not urgent, I think but it surely is a very interesting proposal. Maybe it'd make tagging the source of hash much easier. Let's keep it open for now.
